File: v2.x/Sandbox/SciChart_Boilerplate_ObjC_rolloverLineWithoutInterpolation/SciChart_Boilerplate_ObjC/ChartView.m
//
//  LineChartViewController.m
//  SciChartDemo
//
//  Created by Yaroslav Pelyukh on 1/23/16.
//  Copyright © 2016 SciChart Ltd. All rights reserved.
//

#import "ChartView.h"
#import <SciChart/SciChart.h>
#import "CustomRollover.h"

@implementation ChartView

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    
    if (self) {
        SCIChartSurface * view = [[SCIChartSurface alloc]initWithFrame:frame];
        _surface = view;
        
        [_surface setTranslatesAutoresizingMaskIntoConstraints:NO];
        
        [self addSubview:_surface];
        NSDictionary *layout = @{@"SciChart":_surface};
        
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-(0)-[SciChart]-(0)-|" options:0 metrics:0 views:layout]];
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(0)-[SciChart]-(0)-|" options:0 metrics:0 views:layout]];
        
        [self initializeSurfaceData];
    }
    
    return self;
}

-(instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        SCIChartSurface * view = [[SCIChartSurface alloc]initWithCoder:aDecoder];
        _surface = view;
        
        [_surface setTranslatesAutoresizingMaskIntoConstraints:NO];
        
        [self addSubview:_surface];
        NSDictionary *layout = @{@"SciChart":_surface};
        
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"|-(0)-[SciChart]-(0)-|" options:0 metrics:0 views:layout]];
        [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(0)-[SciChart]-(0)-|" options:0 metrics:0 views:layout]];
        
        [self initializeSurfaceData];
    }
    return self;
}

-(void) initializeSurfaceData {
    [self addAxes];
    [self addModifiers];
    [self addRenderableSeries];
}

-(void) addAxes{
    id<SCIAxis2DProtocol> axis = [[SCINumericAxis alloc] init];
    axis.axisId = @"yAxis";
    [axis setGrowBy: [[SCIDoubleRange alloc]initWithMin:SCIGeneric(0.1) Max:SCIGeneric(0.1)]];
    [_surface.yAxes add:axis];
    
    axis = [[SCINumericAxis alloc] init];
    axis.axisId = @"xAxis";
    [axis setGrowBy: [[SCIDoubleRange alloc]initWithMin:SCIGeneric(0.1) Max:SCIGeneric(0.1)]];
    [_surface.xAxes add:axis];
}

-(void) addModifiers{
    SCIXAxisDragModifier * xDragModifier = [SCIXAxisDragModifier new];
    xDragModifier.axisId = @"xAxis";
    xDragModifier.dragMode = SCIAxisDragMode_Pan;
    xDragModifier.clipModeX = SCIClipMode_None;
    
    SCIYAxisDragModifier * yDragModifier = [SCIYAxisDragModifier new];
    yDragModifier.axisId = @"yAxis";
    yDragModifier.dragMode = SCIAxisDragMode_Pan;
    
    SCIPinchZoomModifier * pzm = [[SCIPinchZoomModifier alloc] init];
    
    SCIZoomExtentsModifier * zem = [[SCIZoomExtentsModifier alloc] init];
    
    CustomRollover * rollover = [CustomRollover new];
    rollover.style.hitTestMode = SCIHitTest_Vertical;
    rollover.hitTestRadius = 50;
    
    SCIChartModifierCollection * gm = [[SCIChartModifierCollection alloc] initWithChildModifiers:@[xDragModifier, yDragModifier, pzm, zem, rollover]];
    _surface.chartModifiers = gm;
}

-(void) addRenderableSeries{
    SCIXyDataSeries * dataSeries = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Float YType:SCIDataType_Float];
    int dataCount = 20;
    for (int i = 0; i < dataCount; i++) {
        double time = 10 * i / (double)dataCount;
        double x = time;
        double y = arc4random_uniform(20);
        [dataSeries appendX:SCIGeneric(x) Y:SCIGeneric(y)];
    }
    
    SCIFastLineRenderableSeries * lineSeries = [SCIFastLineRenderableSeries new];
    [lineSeries setXAxisId: @"xAxis"];
    [lineSeries setYAxisId: @"yAxis"];
    [lineSeries setDataSeries:dataSeries];
    [_surface.renderableSeries add:lineSeries];
    
    [_surface invalidateElement];
}

@end

File: v2.x/Sandbox/SciChart_Boilerplate_Swift_lineAnnotationPositionChanged/SciChart_Boilerplate_Swift/ChartView.swift
//
//  SCSLineChartView.swift
//  SciChartSwiftDemo
//
//  Created by Mykola Hrybeniuk on 5/30/16.
//  Copyright © 2016 SciChart Ltd. All rights reserved.
//

import UIKit
import SciChart

class ChartView: SCIChartSurface {
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        completeConfiguration()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        completeConfiguration()
    }
    
    // MARK: Private Functions

    fileprivate func completeConfiguration() {
        addAxes()
        addModifiers()
        addSeries()
        addAnnotation()
    }
    
    fileprivate func addSeries() {
        let dataSeries = SCIXyDataSeries(xType: .float, yType: .float)
        for i in 0...10 {
             let yValue = Int(arc4random_uniform(10))
            dataSeries.appendX(SCIGeneric(i), y: SCIGeneric(yValue))
        }

        let renderSeries = SCIFastLineRenderableSeries()
        renderSeries.dataSeries = dataSeries
        self.renderableSeries.add(renderSeries)
        
        self.invalidateElement()
    }
    
    fileprivate func addAxes() {
        self.xAxes.add(SCINumericAxis())
        self.yAxes.add(SCINumericAxis())
    }
    
    func addModifiers() {
        let xAxisDragmodifier = SCIXAxisDragModifier()
        xAxisDragmodifier.dragMode = .scale
        xAxisDragmodifier.clipModeX = .none
        
        let yAxisDragmodifier = SCIYAxisDragModifier()
        yAxisDragmodifier.dragMode = .pan
        
        let extendZoomModifier = SCIZoomExtentsModifier()
        
        let pinchZoomModifier = SCIPinchZoomModifier()
        
        let zoomPanModifier = SCIZoomPanModifier()
        
        let groupModifier = SCIChartModifierCollection(childModifiers: [xAxisDragmodifier, yAxisDragmodifier, pinchZoomModifier, extendZoomModifier, zoomPanModifier])
        
        self.chartModifiers = groupModifier
    }
    
    func addAnnotation() {
        let line = CustomLineAnnotation();
        line.coordinateMode = .absolute;
        line.isEditable = true;
        line.x1 = SCIGeneric(1);
        line.y1 = SCIGeneric(1);
        line.x2 = SCIGeneric(7);
        line.y2 = SCIGeneric(5);
        line.lineMoved = { (moved, x1, y1, x2, y2) -> Void in
            if (moved) {
                NSLog("Moved to X1:%g Y1:%g X2:%g Y2:%g", SCIGenericDouble(x1), SCIGenericDouble(y1), SCIGenericDouble(x2), SCIGenericDouble(y2));
            }
        }
        self.annotations.add(line);
    }
    
}
